File: src/main/kotlin/sunsetsatellite/lang/sunlite/Disassembler.kt
package sunsetsatellite.lang.sunlite

import sunsetsatellite.vm.sunlite.Chunk
import sunsetsatellite.vm.sunlite.Opcodes
import sunsetsatellite.vm.sunlite.SLFuncObj

object Disassembler {
	fun disassembleChunk(chunk: Chunk): String {
		val sb = StringBuilder()
		sb.append("==== ${chunk.debugInfo.file.toString()}::${chunk.debugInfo.name} ====\n")

		var offset = 0
		while (offset < chunk.size()) {
			offset = disassembleInstruction(sb, chunk, offset)
		}
		sb.append("==== Exception Table ====\n")

		for ((index, exception) in chunk.exceptions.entries.withIndex()) {
			sb.append(exception.key.toString() + " -> " + exception.value.toString() + "\n")
		}

		sb.append("==== ${chunk.debugInfo.file.toString()}::${chunk.debugInfo.name} ====\n")
		//sb.append("=====${"=".repeat(chunk.debugInfo.file?.length?.plus(chunk.debugInfo.name.length) ?: 0)}=====\n")
		return sb.toString()
	}

	fun disassembleInstruction(sb: StringBuilder, chunk: Chunk, offset: Int): Int {
		sb.append(String.format("%04d ",offset))
		if(offset > 0 && chunk.debugInfo.lines[offset] == chunk.debugInfo.lines[offset - 1]){
			sb.append("   | ");
		} else {
			sb.append(String.format("%4d ",chunk.debugInfo.lines[offset]));
		}
		val inst = chunk.code[offset]
		val opcode: Opcodes = Opcodes.entries[inst.toInt()]
		when (opcode) {
			Opcodes.NOP -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.RETURN -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.CONSTANT -> return constantInstruction(sb, opcode.name, chunk, offset)
			Opcodes.NEGATE -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.ADD -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.SUB -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.MULTIPLY -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.DIVIDE -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.NIL -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.TRUE -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.FALSE -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.NOT -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.EQUAL -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.GREATER -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.LESS -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.PRINT -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.POP -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.DEF_GLOBAL -> return constantInstruction(sb, opcode.name, chunk, offset)
			Opcodes.SET_GLOBAL -> return constantInstruction(sb, opcode.name, chunk, offset)
			Opcodes.GET_GLOBAL -> return constantInstruction(sb, opcode.name, chunk, offset)
			Opcodes.SET_LOCAL -> return shortInstruction(sb, opcode.name, chunk, offset)
			Opcodes.GET_LOCAL -> return shortInstruction(sb, opcode.name, chunk, offset)
			Opcodes.JUMP_IF_FALSE -> return jumpInstruction(sb, opcode.name, 1, chunk, offset)
			Opcodes.JUMP -> return jumpInstruction(sb, opcode.name, 1, chunk, offset)
			Opcodes.LOOP -> return jumpInstruction(sb, opcode.name, -1, chunk, offset)
			Opcodes.CALL -> return twoByteInstruction(sb, opcode.name, chunk, offset)
			Opcodes.CLOSURE -> return closureInstruction(sb, opcode.name, chunk, offset)
			Opcodes.GET_UPVALUE -> return shortInstruction(sb, opcode.name, chunk, offset)
			Opcodes.SET_UPVALUE -> return shortInstruction(sb, opcode.name, chunk, offset)
			Opcodes.CLASS -> return constantInstruction(sb, opcode.name, chunk, offset)
			Opcodes.SET_PROP -> return constantInstruction(sb, opcode.name, chunk, offset)
			Opcodes.GET_PROP -> return constantInstruction(sb, opcode.name, chunk, offset)
			Opcodes.METHOD -> return constantInstruction(sb, opcode.name, chunk, offset)
			Opcodes.FIELD -> return constantInstruction(sb, opcode.name, chunk, offset)
			Opcodes.STATIC_FIELD -> return constantInstruction(sb, opcode.name, chunk, offset)
			Opcodes.INHERIT -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.GET_SUPER -> return constantInstruction(sb, opcode.name, chunk, offset)
			Opcodes.ARRAY_GET -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.ARRAY_SET -> return simpleInstruction(sb, opcode.name, offset)
			Opcodes.THROW -> return simpleInstruction(sb, opcode.name, offset)
            Opcodes.CHECK -> return constantInstruction(sb, opcode.name, chunk, offset)
			Opcodes.TYPE_PARAM -> return constantInstruction(sb, opcode.name, chunk, offset)
        }
	}

	private fun closureInstruction(sb: StringBuilder, name: String, chunk: Chunk, startOffset: Int): Int {
		var offset = startOffset
		val constant = (chunk.code[offset + 1].toInt() shl 8) or chunk.code[offset + 2].toInt()
		offset += 3
		sb.append(String.format("%-16s (%02X) %4d ", name, Opcodes.valueOf(name).ordinal, constant))
		if(constant > chunk.constants.size){
			sb.append("<error reading constant>")
		} else {
			sb.append(chunk.constants[constant].toString())
		}
		sb.append("\n")
		val function = chunk.constants[constant] as SLFuncObj
		for (i in 0 until function.value.upvalueCount) {
			val isLocal = chunk.code[offset++].toInt()
			val index = (chunk.code[offset].toInt() shl 8) or chunk.code[offset + 1].toInt()
			offset += 2
			sb.append(String.format("%04d      L  %s %d\n", offset - 3, if(isLocal == 1) "local" else "upvalue", index))
		}
		return offset
	}

	private fun jumpInstruction(sb: StringBuilder, name: String, sign: Int, chunk: Chunk, offset: Int): Int {
		val jmp = (chunk.code[offset + 1].toInt() shl 8) or chunk.code[offset + 2].toInt()
		sb.append(String.format("%-16s (%02X) %4d -> %d\n", name, Opcodes.valueOf(name).ordinal, offset, offset + 3 + sign * jmp))
		return offset + 3
	}

	private fun byteInstruction(sb: StringBuilder, name: String, chunk: Chunk, offset: Int): Int {
		val byte = chunk.code[offset + 1]
		sb.append(String.format("%-16s (%02X) %4d\n", name, Opcodes.valueOf(name).ordinal, byte))
		return offset + 2
	}

	private fun twoByteInstruction(sb: StringBuilder, name: String, chunk: Chunk, offset: Int): Int {
		val byte = chunk.code[offset + 1]
		val byte2 = chunk.code[offset + 2]
		sb.append(String.format("%-16s (%02X) %4d %4d\n", name, Opcodes.valueOf(name).ordinal, byte, byte2))
		return offset + 2
	}

	private fun shortInstruction(sb: StringBuilder, name: String, chunk: Chunk, offset: Int): Int {
		val short = (chunk.code[offset + 1].toInt() shl 8) or chunk.code[offset + 2].toInt()
		sb.append(String.format("%-16s (%02X) %4d\n", name, Opcodes.valueOf(name).ordinal, short))
		return offset + 3
	}

	private fun constantInstruction(sb: StringBuilder, name: String, chunk: Chunk, offset: Int): Int {
		val constant = (chunk.code[offset + 1].toInt() shl 8) or chunk.code[offset + 2].toInt()
		sb.append(String.format("%-16s (%02X) %4d ", name, Opcodes.valueOf(name).ordinal, constant))
		if(constant > chunk.constants.size){
			sb.append("<error reading constant>")
		} else {
			sb.append(chunk.constants[constant].toString())
		}
		sb.append("\n")
		return offset + 3
	}

	private fun simpleInstruction(sb: StringBuilder, name: String, offset: Int): Int {
		sb.append(String.format("%-16s (%02X)\n", name, Opcodes.valueOf(name).ordinal))
		return offset + 1
	}


}

File: src/main/csharp/SunliteSharp/Runtime/Disassembler.cs
using System.Text;
using SunliteSharp.Util;

namespace SunliteSharp.Runtime;

public static class Disassembler
    {
        public static string DisassembleChunk(Chunk chunk)
        {
            var sb = new StringBuilder();
            sb.Append($"==== {chunk.Header.File}::{chunk.Header.Name} ====\n");

            var offset = 0;
            while (offset < chunk.Length())
            {
                offset = DisassembleInstruction(sb, chunk, offset);
            }

            sb.Append("==== Exception Table ====\n");
            var idx = 0;
            foreach (KeyValuePair<IntRange, IntRange> entry in chunk.Exceptions)
            {
                sb.Append(entry.Key).Append(" -> ").Append(entry.Value).Append('\n');
                idx++;
            }

            sb.Append($"==== {chunk.Header.File}::{chunk.Header.Name} ====\n");
            return sb.ToString();
        }

        public static int DisassembleInstruction(StringBuilder sb, Chunk chunk, int offset)
        {
            sb.Append($"{offset:0000} ");
            if (offset > 0 && chunk.Header.Lines[offset] == chunk.Header.Lines[offset - 1])
            {
                sb.Append("   | ");
            }
            else
            {
                sb.Append($"{chunk.Header.Lines[offset],4} ");
            }

            var inst = chunk.Code[offset];
            var opcode = (Opcodes)inst;

            return opcode switch
            {
                Opcodes.Nop or Opcodes.Return => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.Constant => ConstantInstruction(sb, opcode.ToString(), chunk, offset),
                Opcodes.Negate => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.Add => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.Sub => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.Multiply => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.Divide => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.Nil => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.True => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.False => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.Not => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.Equal => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.Greater => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.Less => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.Pop => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.DefGlobal => ConstantInstruction(sb, opcode.ToString(), chunk, offset),
                Opcodes.SetGlobal => ConstantInstruction(sb, opcode.ToString(), chunk, offset),
                Opcodes.GetGlobal => ConstantInstruction(sb, opcode.ToString(), chunk, offset),
                Opcodes.SetLocal or Opcodes.GetLocal => ShortInstruction(sb, opcode.ToString(), chunk, offset),
                Opcodes.JumpIfFalse or Opcodes.Jump => JumpInstruction(sb, opcode.ToString(), 1, chunk, offset),
                Opcodes.Loop => JumpInstruction(sb, opcode.ToString(), -1, chunk, offset),
                Opcodes.Call => TwoByteInstruction(sb, opcode.ToString(), chunk, offset),
                Opcodes.Closure => ClosureInstruction(sb, opcode.ToString(), chunk, offset),
                Opcodes.GetUpvalue or Opcodes.SetUpvalue => ShortInstruction(sb, opcode.ToString(), chunk, offset),
                Opcodes.Class => ConstantInstruction(sb, opcode.ToString(), chunk, offset),
                Opcodes.SetProp => ConstantInstruction(sb, opcode.ToString(), chunk, offset),
                Opcodes.GetProp => ConstantInstruction(sb, opcode.ToString(), chunk, offset),
                Opcodes.Method => ConstantInstruction(sb, opcode.ToString(), chunk, offset),
                Opcodes.Field => ConstantInstruction(sb, opcode.ToString(), chunk, offset),
                Opcodes.StaticField => ConstantInstruction(sb, opcode.ToString(), chunk, offset),
                Opcodes.Inherit => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.GetSuper => ConstantInstruction(sb, opcode.ToString(), chunk, offset),
                Opcodes.ArrayGet => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.ArraySet => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.Throw => SimpleInstruction(sb, opcode.ToString(), offset),
                Opcodes.Check or Opcodes.TypeParam => ConstantInstruction(sb, opcode.ToString(), chunk, offset),
                _ => throw new ArgumentOutOfRangeException(nameof(offset),"Invalid opcode")
            };
        }

        private static int ClosureInstruction(StringBuilder sb, string name, Chunk chunk, int startOffset)
        {
            var offset = startOffset;
            var constant = (chunk.Code[offset + 1] << 8) | chunk.Code[offset + 2];
            offset += 3;
            sb.Append($"{name,-16} ({(int)Enum.Parse<Opcodes>(name):X2}) {constant,4} ");
            sb.Append(constant > chunk.Constants.Length
                ? "<error reading constant>"
                : chunk.Constants[constant]?.ToString());
            sb.Append('\n');

            var function = (SLFuncObj)chunk.Constants[constant];
            for (var i = 0; i < function.Value.UpvalueCount; i++)
            {
                int isLocal = chunk.Code[offset++];
                var index = (chunk.Code[offset] << 8) | chunk.Code[offset + 1];
                offset += 2;
                sb.Append($"{offset - 3:0000}      L  {((isLocal == 1) ? "local" : "upvalue")} {index}\n");
            }
            return offset;
        }

        private static int JumpInstruction(StringBuilder sb, string name, int sign, Chunk chunk, int offset)
        {
            var jmp = (chunk.Code[offset + 1] << 8) | chunk.Code[offset + 2];
            sb.Append(
                $"{name,-16} ({(int)Enum.Parse<Opcodes>(name):X2}) {offset,4} -> {offset + 3 + sign * jmp}\n");
            return offset + 3;
        }

        private static int ByteInstruction(StringBuilder sb, string name, Chunk chunk, int offset)
        {
            var b = chunk.Code[offset + 1];
            sb.Append($"{name,-16} ({(int)Enum.Parse<Opcodes>(name):X2}) {b,4}\n");
            return offset + 2;
        }

        private static int TwoByteInstruction(StringBuilder sb, string name, Chunk chunk, int offset)
        {
            var b1 = chunk.Code[offset + 1];
            var b2 = chunk.Code[offset + 2];
            sb.Append($"{name,-16} ({(int)Enum.Parse<Opcodes>(name):X2}) {b1,4} {b2,4}\n");
            return offset + 2;
        }

        private static int ShortInstruction(StringBuilder sb, string name, Chunk chunk, int offset)
        {
            var val = (chunk.Code[offset + 1] << 8) | chunk.Code[offset + 2];
            sb.Append($"{name,-16} ({(int)Enum.Parse<Opcodes>(name):X2}) {val,4}\n");
            return offset + 3;
        }

        private static int ConstantInstruction(StringBuilder sb, string name, Chunk chunk, int offset)
        {
            var constant = (chunk.Code[offset + 1] << 8) | chunk.Code[offset + 2];
            sb.Append($"{name,-16} ({(int)Enum.Parse<Opcodes>(name):X2}) {constant,4} ");
            sb.Append(constant > chunk.Constants.Length
                ? "<error reading constant>"
                : chunk.Constants[constant]?.ToString());
            sb.Append('\n');
            return offset + 3;
        }

        private static int SimpleInstruction(StringBuilder sb, string name, int offset)
        {
            sb.Append($"{name,-16} ({(int)Enum.Parse<Opcodes>(name):X2})\n");
            return offset + 1;
        }
    }